### PR TITLE
Project updated when subproject modified

### LINKED ISSUE #3396: Subproject is not updated on save/build
When changing a subproject of a superproject, for instance updating the privacy level of the project, symlinks for the superproject are not updated. This might be a regression.

## To reproduce

* Create a project A, privacy level public
* Create a project B, privacy level private
* Assign project B as a subproject of project A
* Project B URLs should 404
* Alter project B privacy level to public
* Project B URLs still 404

## Expected Result

* Project B URLs shouldn't 404 after updating to public privacy level

## Resolving

We should either trigger a save of the project's superproject, or explicitly call a task for re-symlinking the project in the following cases:

* When a subproject is deleted
* When a subproject privacy level is changed
* When a subproject version privacy level is changed (?? not sure here, might need testing on this as well)

### PR DESCRIPTION
**Pull request by humitos**
Closes #3396

To Do:

- [x] find the correct before/after filesystem
- [x] write the test in a proper way
- [x] make the test fail/pass manually
- [x] implement the solution proposed in the original issue

### CODE REVIEW

**Comment by humitos** [comment]
I have some questions/doubts that are related to this PR (maybe to be implemented in another -but related):

* when a Version is deleted, we are triggering `clear_artifacts` _and_ `symlink_project` but the later is triggered _before_ object deletion; so, I think if the task is ran before it's effectively deleted the symlink won't be properly synced:

https://github.com/rtfd/readthedocs.org/blob/5472906a067adc60f42556108993255393408410/readthedocs/builds/models.py#L181-L187

* when a (sub)Project is deleted we don't sync anything. I think that maybe we should sync the superprojects

* when a (super)Project is deleted, what should happen with the subprojects?

@agjohnson the PR for the original issue is ready to review --let me know if you want me to squash the commits into one. Although, I would like you to take a look a those questions that I'm raising here.

**Comment by ericholscher** [comment]
There are a lot of bugs in our "keep symlinks up to date" code. We should write a design document that specifies the exact place that we _should_ be doing sync'ing for this, and then do an audit to make sure that we're actually doing it.

I've tried to build a proper abstraction for maintaining the symlinks in code, but that never got pushed back into a proper abstraction for updating/removing/editing symlinks with the lifecycle of the model's. I think that is what this PR (and your questions) are addressing. 

What is the *correct* way to handle updating & syncing symlinks in all our code, is the question we need to answer.

File: requirements/testing.txt [hunk 1]
```diff
@@ -9,3 +9,4 @@ Mercurial==4.4.2
 
 # local debugging tools
 pdbpp
+datadiff
```
Comment (agjohnson) [inline requirements/testing.txt]: looks unused, but is it helpful to keep around?

**Comment by agjohnson** [comment]
A design document might be good to have in the future, but i think will slow down this work without much gain to be had. I'll create a new issue for this under our refactoring milestone.

**Comment by humitos** [comment]
@agjohnson I just pushed a change with a test to check that `broadcast` function was called as it's expected. Let me know if that is what you were thinking.

File: requirements/testing.txt (hunk 1 above)
Comment (agjohnson) [reply requirements/testing.txt]: great, let's keep it for debug purposes

**Comment by agjohnson** [comment]
Looks great!

**Pull request merged**